torchvision/datapoints/_dataset_wrapper.py

```diff
@@ -16,7 +15,7 @@
 
 # TODO: naming!
 def wrap_dataset_for_transforms_v2(dataset):
-    return VisionDatasetDatapointWrapper(dataset)
+    return type(f"Wrapped{type(dataset).__name__}", (VisionDatasetDatapointWrapper, type(dataset)), {})(dataset)
```

Comment:
> OK so we're creating a class on the fly named `WrappedImageNet` that inherits from both `VisionDatasetDatapointWrapper` _and_ `ImageNet`, and we instanciate it with `ImageNet()` which is in fact passed to the constructor of `VisionDatasetDatapointWrapper`?

Correct. I think that is what you meant, but the `ImageNet()` instance is what we get from the user. Meaning, we never actually create a dataset object inside the wrapper.

> So `isinstance(ds, ImageNet)` and `isinstance(ds, VisionDatasetDatapointWrapper)` and `isinstance(ds, WrappedImageNet)` are all true (except that `WrappedImageNet` isn't in scope anywhere)?

Correct.

> (except that `WrappedImageNet` isn't in scope anywhere)

Well that is what users get when they do `type(wrapped_dataset)`, but I'm ok with that since it is not really different from before where this returned `VisionDatasetDatapointWrapper`. This also is never exposed anywhere. Meaning, if users only use `isinstace` we are good. If they go near `type` it will be complicated.

> I assume that this doesn't work:
> 
> ```python
> type(f"Wrapped{type(dataset).__name__}", (type(dataset),), {})(VisionDatasetDatapointWrapper(dataset))
> ```
> 
> because the `ImageNet` constructor doesn't expect a `VisionDatasetDatapointWrapper()` instance?

Correct. If we inherit from `ImageNet` and don't define a constructor, the one from the superclass is called.